File: DP/boj1509_lsh.py
import math, sys
# sys.setrecursionlimit(10000)


def is_palindrome(s):
    left, right = 0, len(s) - 1
    while left < right:
        if s[left] != s[right]:
            return False
        left += 1
        right -= 1
    return True

def recur(left, right):
    if cache[left][right] != -1:
        return cache[left][right]
    if left >= right:
        return 1
    if left + 1 == right:
        if is_palindrome(s[left:right + 1]):
            return 1
        else:
            return 2

    if is_palindrome(s[left:right + 1]):
        cache[left][right] = 1
        return cache[left][right]

    ret = math.inf
    for middle in range(left + 1, right + 1):
        ret = min(ret, recur(left, middle - 1) + recur(middle, right))
    cache[left][right] = ret

    return cache[left][right]



s = input()
cache = [[-1] * (len(s) + 1) for _ in range(len(s) + 1)]
print(recur(0, len(s) - 1))

File: DP/boj1509_lsh.cpp
#include <iostream>
#include <string>
#include <memory.h>

using namespace std;

int cache[2501][2501];
int palin[2501][2501];
string s;

bool isPalindrome(int left, int right)
{
  int &ref = palin[left][right];

  if (ref != -1)
    return ref;

  while (left < right)
  {
    if (s[left] != s[right])
      return ref = 0;

    left++;
    right--;
  }

  return ref = 1;
}

int recur(int left, int right)
{
  int &ref = cache[left][right];

  if (ref != -1)
    return ref;

  if (left >= right)
    return 1;
  if (left + 1 == right)
  {
    if (isPalindrome(left, right))
      return 1;
    else
      return 2;
  }

  if (isPalindrome(left, right))
    return ref = 1;

  int ret = 500000;

  for (int middle = left; right >= middle; middle++)
  {
    if (isPalindrome(left, middle))
      ret = min(ret, recur(middle + 1, right) + 1);
  }
  return ref = ret;
}

int main(void)
{
  cin >> s;
  memset(cache, -1, sizeof(cache));
  memset(palin, -1, sizeof(palin));

  cout << recur(0, s.size() - 1) << '\n';
}